### PR TITLE
#174250315 add payment plan

### PR DESCRIPTION
**Pull request by tobieeniafe**
#### What does this PR do?
Add payment plan endpoints

#### Description of Task to be completed?
add endpoints for admin to create, view and delete payment plans

#### How should this be manually tested?
`POST /api/v1/payment-plan/add`
`{
  "planName": "monthly",
  "planDescription": "monthly payment plan",
  "paymentFrequency": 1
}`
`GET /api/v1/payment-plan/all`
`DELETE /api/v1/payment-plan/delete/:id`

#### What are the relevant pivotal tracker stories?
#174250315

### CODE REVIEW

File: server/controllers/paymentPlan.controllers.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    const newPaymentPlan = req.locals;
+    const { user } = req;
+    addPaymentPlan({ ...newPaymentPlan, addedBy: user._id })
+      .then((plan) => {
+        res.status(httpStatus.CREATED).json({ success: true, message: 'Plan added', plan });
+      })
+      .catch((error) => next(error));
+  },
+
+  delete(req, res, next) {
```
Comment (andela-hpopoola) [inline server/controllers/paymentPlan.controllers.js:19]: 🤔 . This is a dicey situation. We should be careful with the delete feature with everything that deals with money. The delete endpoint for the payment plan should only delete the payment plan if it is not used by any property. A deleted payment plan in use will result in a bug if the property is retrieved.

File: server/models/paymentPlan.model.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  {
+    planName: {
+      type: String,
+      required: true,
+    },
+    planDescription: {
+      type: String,
+      required: true,
+    },
+    paymentFrequency: {
```
Comment (andela-hpopoola) [inline server/models/paymentPlan.model.js:37]: This should be a number that denotes month. It will be difficult to automate calculations from payment plans if this is a string.

File: server/models/paymentPlan.model.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ *        example:
+ *           planName: Weekly payment
+ *           planDescription: payment is made once a week
+ *           paymentFrequency: weekly
+ */
+
+const { ObjectId } = mongoose.Schema.Types;
+const PaymentPlanSchema = new mongoose.Schema(
+  {
+    planName: {
```
Comment (andela-hpopoola) [inline server/models/paymentPlan.model.js:29]: We could as well use name here (the plan prefix is not necessary, I notice that you do that for most of your naming convention). 

This will make the code more readable, something like `paymentPlan.name` seems better than `paymentPlan.planName`.

File: server/routes/index.js [hunk 4]
```diff
@@ -5,6 +5,7 @@ import property from './property.routes';
 import enquiry from './enquiry.routes';
 import mailer from './mailer.routes';
 import visitation from './visitation.routes';
+import payment from './paymentPlan.routes';
```
Comment (andela-hpopoola) [inline server/routes/index.js:8]: Lets be explicit with paymentPlan as we don't want a mix up with the payment and transaction model.

```suggestion
import paymentPlan from './paymentPlan.routes';
```

File: server/routes/index.js [hunk 5]
```diff
@@ -13,6 +14,7 @@ export default (app) => {
   app.use('/api/v1/property', property);
   app.use('/api/v1/enquiry', enquiry);
   app.use('/api/v1/visitation', visitation);
+  app.use('/api/v1/payment', payment);
```
Comment (andela-hpopoola) [inline server/routes/index.js:17]:
```suggestion
  app.use('/api/v1/payment-plan', paymentPlan);
```

File: server/routes/paymentPlan.routes.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
+import express from 'express';
+import { authenticate, schemaValidation, isAdmin, hasValidObjectId } from '../helpers/middleware';
+import addPlanSchema from '../schemas/paymentPlan.schema';
```
Comment (andela-hpopoola) [inline server/routes/paymentPlan.routes.js:3]: `addPlanSchema` is not explicit enough. The codebase is getting bigger, it is easier to mix things up now.
```suggestion
import addPaymentPlanSchema from '../schemas/paymentPlan.schema';
```

File: server/routes/paymentPlan.routes.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+import PaymentPlanController from '../controllers/paymentPlan.controllers';
+
+const router = express.Router();
+
+/**
+ * @swagger
+ * /payment/add:
```
Comment (andela-hpopoola) [inline server/routes/paymentPlan.routes.js:10]: Please fix for other endpoints too.
```suggestion
 * /payment-plan/add:
```

File: server/routes/paymentPlan.routes.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+ *            $ref: '#/components/schemas/PaymentPlan'
+ *      description: Get all owned enquiries
+ *     responses:
+ *      '200':
+ *        description: returns object of payment plans
+ *      '500':
+ *       description: Internal server error
+ */
+router.get('/all', authenticate, isAdmin, PaymentPlanController.getAll);
+
```
Comment (andela-hpopoola) [inline server/routes/paymentPlan.routes.js:61]: We should have an endpoint to edit payment plans. Editing is important for fixing typos in plan name and making the description more descriptive.

However, the frequency field should not be updated since updating this will affect all the previously calculated price made with this payment plans.

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for working on this. 

Can you integrate the payment plan model into the property model so that when properties are loaded via any of the property endpoint, the payment plan is also returned among the property details. This is the most important reason why this is important in the added to MVP. Thank you.

File: server/models/property.model.js [hunk 9]
```diff
@@ -129,6 +129,9 @@ const PropertySchema = new mongoose.Schema(
     assignedTo: {
       type: [ObjectId],
     },
+    paymentPlan: {
```
Comment (andela-hpopoola) [inline server/models/property.model.js:132]: PaymentPlan should be an array of ObjectId. 
(If you study the mockup - you will notice that we have multiple payment plans.)

Your implementation is different from the concept in the mockup 😏  

In the mockup, you need to study all the places the payment plan is used, you will notice that it is returned with the property information.

1. Single Portfolio Page
 This means that the returned result for the `/property/:id` endpoint need to return the names of the payment plan.

Screenshot 2020-08-25 at 07 13 04


2. List of all Properties contains payment plan. We need to ensure that the payment plan is aggregated in all property endpoint so that it is returned as part of the property information.

See another instance here:
Screenshot 2020-08-25 at 07 23 28

To save time and energy, lets ensure we return the aggregated payment plan for all property endpoints.

File: server/controllers/paymentPlan.controllers.js [hunk 10]
```diff
@@ -32,6 +34,27 @@ const PaymentPlanController = {
...
+    const updatedPaymentPlan = req.locals;
+    const { user } = req;
+    updatePaymentPlan({ ...updatedPaymentPlan, updatedBy: user._id })
+      .then((plan) => {
+        res.status(httpStatus.OK).json({ success: true, message: 'Payment plan updated', plan });
+      })
+      .catch((error) => next(error));
+  },
+
+  assignPaymentPlan(req, res, next) {
```
Comment (andela-hpopoola) [inline server/controllers/paymentPlan.controllers.js:48]: This is a great implementation, however, it will cause the user and application bad user experience. I believe payment plans should be added when the user (developer/admin) is adding a new property (not assigned individually). This will be implemented via a multiple Select field in the frontend so that the user can add and edit payment plans at ease.

A better implementation will be to remove this method and add payment plans to add property/update property implementation. The ids of the payment plan are saved in the property model as an array. If you are curious about the multi select, here is a link to the one I used in the application https://react-select.com/home (I am referring to the 2nd example for multiple select field)

I hope you understand, please free to follow up with me if this is unclear or you have a much better implementation in mind.

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for working on this. I believe the implementation can be improved to make the endpoint more effective.

File: server/models/paymentPlan.model.js [hunk 11]
```diff
@@ -7,41 +7,47 @@ import mongoose from 'mongoose';
...
+    description: {
       type: String,
       required: true,
     },
     paymentFrequency: {
-      type: String,
+      type: Number,
       required: true,
     },
+    propertiesAssignedTo: {
```
[the remark this answers is not in the document]
Comment (tobieeniafe) [reply server/models/paymentPlan.model.js:41]: this is to kow is a payment plan is linked to a property, to prevent it from being deleted

[the remark this answers is not in the document]
Comment (andela-hpopoola) [reply server/models/paymentPlan.model.js:41]: I understand your point but this is an extra overhead over the payment plan and it is not necessary(as we will need to add an extra query to add the properties to the payment plan). This is extra work as we will need to keep track of the payment plans in the property model and also keep track of the payment plan model at the same time. If the payment plan field in the property is added, we must add the property field too in the payment plan model. This also applies to when a property is updated and deleted. This will be extra code and extra work (it doesn't worth the stress, abeg)

Moreover, we can know if a payment plan is linked to a property by querying the properties since the payment plan is saved in the property model too. 🤷‍♂️

[the remark this answers is not in the document]
Comment (tobieeniafe) [reply server/models/paymentPlan.model.js:41]: `hasBeenAssignedToProperty` `true` ?🤔

[the remark this answers is not in the document]
Comment (andela-hpopoola) [reply server/models/paymentPlan.model.js:41]: The reason why querying the property model is the best approach is because it will only be done when the admin is deleting. The problem with this are:
1. You will need to write an additional query to update the payment plan `hasBeenAssignedToProperty` field to true every time a new property is added. 
2. We are not assigning payment plans one by one via an endpoint. You will need to remove the assignment endpoint, We are assigning payment plans is multiples -- An admin can assign 4 to 5 payment plan at the same time when he is adding a new property (remember the multi-select field in my earlier comment), that means we will make 4 to 5 queries to update the respective payment plan `hasBeenAssignedToProperty`.
3. Every time a property is updated, we need to check that the payment plan has not been updated because if it has been updated (i.e a payment plan has been removed, and another one has been added) we will need to check if the removed property is the only property with the payment plan and use that to update the value of `hasBeenAssignedToProperty`.

I am actually being considerate by telling you to remove this. However, if you think adding the `asBeenAssignedToProperty` field approach is better, my only plead is for you to write tests to at least simulate my points 2 and 3 so that we are certain that we cover all the edge cases.

File: server/schemas/paymentPlan.schema.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+  paymentFrequency: Joi.number().label('Plan Payment Frequency').required(),
+});
+
+export const updatePaymentPlanSchema = Joi.object({
+  id: Joi.objectId().label('Plan ID').required(),
+  name: Joi.string().label('Plan Name').optional(),
+  description: Joi.string().label('Plan Description').optional(),
+});
+
+export const assignPaymentPlanSchema = Joi.object({
```
Comment (andela-hpopoola) [inline server/schemas/paymentPlan.schema.js:17]: Can we remove this, not sure we need it or what do you think?

File: server/services/paymentPlan.service.js [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+export const deletePaymentPlan = async (id) => {
+  const plan = await getPaymentPlanById(id).catch((error) => {
+    throw new ErrorHandler(httpStatus.INTERNAL_SERVER_ERROR, 'Internal Server Error', error);
+  });
+  const propertiesAssignedTo = await getPropertiesWithPaymentPlanId(id);
+  if (propertiesAssignedTo.length > 0) {
+    throw new ErrorHandler(
+      httpStatus.PRECONDITION_FAILED,
+      'Cannot delete payment plan with properties assigned',
```
Comment (andela-hpopoola) [inline server/services/paymentPlan.service.js:25]: Not too important but necessary for User Experience, I believe that the error message will be more informative if we return the name of the properties (as an array) the payment plan is currently assigned to. It will make it easier to locate the property and update it.

Comment (andela-hpopoola) [inline server/services/paymentPlan.service.js:25]: It is advisable to make the error message more descriptive. Let's add the number of properties it is assigned to. Please use `property` for 1 property and plural form for more than one property in the updated error message as shown below:

```suggestion
      `Payment plan cannot be deleted as it is currently assigned to 2 properties`,
```

**Review by andela-hpopoola — APPROVED** [review]
Thank you for working on this. I left a few non blocking comments. Feel to free to merge it after fixing them.

File: server/services/paymentPlan.service.js (hunk 13 above)
Comment (tobieeniafe) [reply server/services/paymentPlan.service.js]: any suggestion on how to implement this?
